artemis-commons/src/main/java/org/apache/activemq/artemis/utils/critical/CriticalMeasure.java

```diff
@@ -17,28 +17,39 @@
 
 package org.apache.activemq.artemis.utils.critical;
 
-import org.jboss.logging.Logger;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
 
-public class CriticalMeasure {
+final class CriticalMeasure {
 
-   private static final Logger logger = Logger.getLogger(CriticalMeasure.class);
+   //uses updaters to avoid creates many AtomicLong instances
+   private static final AtomicLongFieldUpdater<CriticalMeasure> TIME_ENTER_UPDATER = AtomicLongFieldUpdater.newUpdater(CriticalMeasure.class, "timeEnter");
+   private static final AtomicLongFieldUpdater<CriticalMeasure> TIME_LEFT_UPDATER = AtomicLongFieldUpdater.newUpdater(CriticalMeasure.class, "timeLeft");
 
-   private volatile long timeEnter;
-   private volatile long timeLeft;
+   //System::nanoTime can't reach this value so it's the best candidate to have a NULL semantic
+   private static final long NIL = Long.MAX_VALUE;
+   private volatile long timeEnter = NIL;
+   private volatile long timeLeft = NIL;
 
    public void enterCritical() {
-      timeEnter = System.currentTimeMillis();
+      //prefer lazySet in order to avoid heavy-weight full barriers
+      TIME_ENTER_UPDATER.lazySet(this, System.nanoTime());
    }
 
    public void leaveCritical() {
-      timeLeft = System.currentTimeMillis();
+      final long now = System.nanoTime();
+      assert TIME_ENTER_UPDATER.get(this) <= now : "Must call enterCritical first";
```

Comment:
you're adding a risky semantic check here for no reason.. just do what the previous version was doing please?

Reply:
+1